### ISSUE TITLE
Memory usage spikes uncontrollably

### ISSUE THREAD

**Issue by Moondownr** [slskd/slskd#251]
I've been using slskd for 3 days on my home server, running Ubuntu 20.04 with 4gb ram. I'm sharing ~1TB worth of files: mainly music, some pdfs. The server also has NFS and SMB shares on same location that slskd is pointed at. 

What I noticed is that slskd normally uses 150-200 mb of ram, but once in 24-36 hours something causes it to start consuming more, to the point of 2GB+, at which point my server runs out of memory. I only modified paths in the config. I'm not using docker, I set up a systemd service to run slskd. The underlying file system is btrfs.

What I did:
- move files in and out of folders that slskd shares, without rescanning/restarting slskd.
- move/delete files out of slskd download folders
...basically, nothing else. If there are any logs I can provide, I'll be glad to. Unfortunately, I cannot predict when the problem occurs or what causes it.

**Comment by jpdillingham** [comment]
It would be helpful if you could zip what logs you have currently and email them to `logs@slskd.org` (to protect your privacy, I'd rather you didn't post them publicly anywhere, as they will reveal files you are sharing)

I would venture a guess that someone is enqueueing a bunch of files, as there's currently no controls over how many concurrent uploads can be in progress.  If there's nothing obvious in the logs you provide I can make some changes to help track transfers in progress at any given point.

Depending on your own activities (number of downloads, etc) 2GB might not be an unreasonable amount of memory,  Are you downloading a lot of files when this happens?

**Comment by Moondownr** [comment]
Logs sent to email.

Usually I do not download anything at all when it happens, actually. 
As for uploads, I have been noticing numerous search request errors in logs ("not in dictionary"), which prompted me to do a quick search issues here and find the closed issue re: cache not being filled yet.
Thing is, those keep happening to me even 8 hours and more after scanning shares and I'm not sure how long it might take to fill the cache, although uploads do start happening, it's just that errors don't stop. Most likely an unrelated issue, but just in case/

**Comment by jpdillingham** [comment]
I've created an issue for the search failures in #252; the root cause appears to be single quotes in the filename.  I also noticed a bug in the version check which I've filed in #253, but that only happens one time on start so that shouldn't be the issue.

No logs jump out at me as being abnormal, but I do see that you are cancelling a handful of uploads.  Are a lot of files queued?

I reviewed the logs on my always-on instance and it does appear as though something is killing the `slskd` process once in a while, it's just being restarted automatically so I never noticed.

I'm guessing this is going to be an issue with how search requests are resolved, but I'll see what I can do to add some logging.

On the off chance that you run an instance of Prometheus anywhere in your network, you can scrape metrics from slskd.  Historical data would help.

**Comment by Moondownr** [comment]
I am only cancelling uploads that are red (Errored, Rejected, etc.), because there are a lot and they a cluttering the view. They seem to come back though, sometimes multiple times, after I cleared them. They just reappear and instantly go red again.

No Prometheus, sorry.

**Comment by jpdillingham** [comment]
It's normal for other clients to attempt to restart cancelled downloads, so no problem there.

Can you tell me more about your hardware? CPU, total memory, etc.  Are you sharing any files located on a network drive, or is it all local?

**Comment by jpdillingham** [comment]
I'm playing with ways to try to set an upper limit on the heap size per: https://docs.microsoft.com/en-us/dotnet/core/run-time-config/garbage-collector#heap-limit

So far I've gotten the application to crash when I set `COMPlus_GCHeapHardLimit=6400000` (100mb), so it's definitely doing something.  I've got it set to 200mb now and the metrics are holding steady.  I'll let it run overnight and see what it does.

In the meantime if you could set the same environment variable on your system and report back it would be helpful.

**Comment by mathiascode** [comment]
> I am only cancelling uploads that are red (Errored, Rejected, etc.), because there are a lot and they a cluttering the view. They seem to come back though, sometimes multiple times, after I cleared them. They just reappear and instantly go red again.

Unrelated to this issue, but do the file paths of these failed uploads contain special characters (accented characters etc)?

**Comment by Moondownr** [comment]
> It's normal for other clients to attempt to restart cancelled downloads, so no problem there.

Just to clarify: I do not cancel anything ever, unless it's cancelled automatically - then I try to remove it.
 
> Can you tell me more about your hardware? CPU, total memory, etc. Are you sharing any files located on a network drive, or is it all local?

It's an upgraded 2009 Mac Mini model 3.1: Core 2 Duo, 4gb ram, 2*2TB internal storage (no raid). Running Ubuntu 20.04 server + lxqt de. Everything I share is on an internal hdd. Those same folders that I'm sharing are also being exported as nfs shares and serve as part of the plex media server library.

> Unrelated to this issue, but do the file paths of these failed uploads contain special characters (accented characters etc)?

Some of them, not all of them. It's generally per user, not per specific file - someone queues one or several folders and they all error out.

Also, a lot of files have single quotes in their names, as it's pretty common is song/album/artist titles, and I see many of them uploaded normally (screenshot for reference: https://imgur.com/a/fBNS9YO )

> In the meantime if you could set the same environment variable on your system and report back it would be helpful.

I exported COMPlus_GCHeapHardLimit=12800000

**Comment by jpdillingham** [comment]
Thanks for the additional context regarding the hardware.

The failed uploads, especially those marked as `Rejected` point to a problem with encoding, which I think is what @mathiascode was hinting at.  Errors related to not being able to find files is a problem internal to slskd related to the shared file cache, so unfortunately there's multiple things going on here.

I've been poking and prodding at the application this morning and have managed to get it into a bad state, but I'm not quite sure how or why.  The good news is that the `COMPlus_GCHeapHardLimit` is definitely capping the memory usage and preventing it from bringing down the host.

**Comment by jpdillingham** [comment]
At this point, I'm just going to have to let the application run and see how it behaves over time.  I've got Prometheus scraping metrics from a version running in docker, and will add another running on a resource constrained bare metal VM this weekend to see if that matters.

A few key takeaways:

* The application _does not_ like being memory constrained
* The application uses a lot of memory for some operations, and generally not much at other times
* Memory constraints become apparent only during high demand times, as opposed to a memory leak (pending confirmation)

This is all based on previous diagnostics (I've spent a lot of time eliminating memory leaks in the past) and observations since this issue was raised, but could easily change over days or weeks.

For now, I'm going to recommend that anyone experiencing memory issues sets the following environment variable:

```
COMPlus_GCHeapHardLimit=1F400000
```

`1F400000` is 524288000 or 500 mebibytes in hexadecimal.  I have found 100mb to be totally unusable and 200mb to be generally unstable, so 500mb is a good shot in the dark.  If you aren't sharing many files, or don't use the application much, you can probably get away with less.  If you share a ton of stuff and are very active, you might want to bump the limit to `3D090000` or 1gigabyte.

I'll leave this issue open and will continue to provide updates when I have them.

**Comment by jpdillingham** [comment]
I've been profiling memory literally all day and I believe I may have it narrowed down to the distributed network.

@Moondownr if you would, please set `SLSKD_SLSK_DNET_CHILDREN=1` or (`soulseek.distributed_network.child_limit` if you're using the yaml file).  Please also unset `COMPlus_GCHeapHardLimit`

I haven't been able to prove or disprove this yet, but I believe the following code from the Soulseek.NET library is likely to blame:

```c#
        /// <summary>
        ///     Asynchronously writes the specified bytes to each of the connected child connections.
        /// </summary>
        /// <param name="bytes">The bytes to write.</param>
        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
        /// <returns>The operation context.</returns>
        public async Task BroadcastMessageAsync(byte[] bytes, CancellationToken? cancellationToken = null)
        {
            var tasks = ChildConnectionDictionary.Values.Select(async c =>
            {
                IMessageConnection connection = null;

                try
                {
                    connection = await c.Value.ConfigureAwait(false);
                    await connection.WriteAsync(bytes, cancellationToken ?? CancellationToken.None).ConfigureAwait(false);
                }
                catch (Exception)
                {
                    connection?.Dispose();
                }
            });

            await Task.WhenAll(tasks).ConfigureAwait(false);
        }
```

This is executed each time a search request is received from the server or distributed parent, which is extremely often.  Picking through GC dumps, I kept seeing `Lazy<Task<IMessageConnection>>>` pop up and realized that the `.Select()` in this code is creating a new list every time this runs.

I'm going to refactor this code regardless, but I'm pretty sure it is the culprit.

**Comment by Moondownr** [comment]
> @Moondownr if you would, please set SLSKD_SLSK_DNET_CHILDREN=1 or (soulseek.distributed_network.child_limit if you're using the yaml file). Please also unset COMPlus_GCHeapHardLimit

Done!

(btw, didn't experience any isssues for the last two days with COMPlus_GCHeapHardLimit=1F400000)

**Comment by jpdillingham** [comment]
I've tracked (I hope!) the problem down to distributed child connections failing to read the data that has been sent, particularly distributed search requests.  Here's a graph:

![image](https://user-images.githubusercontent.com/17145758/141992419-9f2057a1-6a37-4fe8-b633-c6897832351c.png)

Each of these spikes correlated to a distributed child connection disconnecting abnormally, each at the instant when the memory usage spiked (broke from the linear climb), and 30 seconds to 1 minute later, memory usage returned to normal.

I suspected this was a problem in the broadcast logic but the issue is one layer lower, in the connection logic itself.  In these cases the socket write buffer increases along with the internal task queue, pushing memory higher at the rate at which we process distributed messages.

The solution is to track how many concurrent connection writes are under way, and once a threshold is broken, to disconnect the connection.  I had originally planned to reserve this "double buffering" to distributed writes only, but it makes sense to apply it to all writes, and set the default queue limit to 500, which seems more than generous.

Writes other than those to a distributed child should never come anywhere near to 500 concurrently.

**Comment by jpdillingham** [comment]
To add additional rationale (I'm linking this issue in code so I can review later, after I've forgotten all of this), here's a graph of the past few days, which me intentionally trying to create memory leaks:

![image](https://user-images.githubusercontent.com/17145758/141995462-1fe87982-77af-46f6-96f5-62bb505bcb1b.png)

I have been tweaking the connection code in Soulseek.NET and trying different approaches.  The 3 spikes on this graph are the ones shown in the post above, and they occurred during a period where double-buffered writes were disabled.

Reviewing logs, there have been 18 instances in the last 2 days where a child was disconnected due to a full buffer.  This is fairly consistent with the rate at which I have been seeing spikes (+/- quite a bit, but it's generally in line).

**Comment by jpdillingham** [comment]
I'm calling this one resolved in the next (0.14) release; I've done a ton of benchmarking and tweaking, and the memory usage is now very stable.  Here's the graph for the last 7 days:

![image](https://user-images.githubusercontent.com/17145758/144615454-c31c45e8-c678-40a8-b5c6-6f1cf0096a44.png)

The working set still tends to climb when a really broad search is done, and won't release the memory very quickly.  I've created #288 to track a feature that will cause a garbage collection after searches, which will force it to release memory quicker.

Memory usage during very broad searches (for example, the spike in the graph is from me searching "mp3") is always going to be high because of the amount of work the application does to receive and process search results.  My guidance is to not do broad searches like this on memory constrained systems.  I'm aware that's sub-optimal, but the only thing I could possibly do to make this better is to outright reject some results, and that's objectively worse.

**Comment by jpdillingham** [comment]
It should be noted that the root cause of the memory spikes was indeed the send buffer of bad child connections filling up.  I've added a sort of double-buffer which will disconnect these connections when the number of enqueued writes goes above the threshold specified in the `write_buffer` option.  In a perfect world these connections would be better about disconnecting themselves but I couldn't find a reliable way to make it happen.

The child limit with the graph above was set to 100, so at any point in the previous 7 days this client has had between 0 and 100 children and the memory has remained flat.

It shouldn't be necessary for anyone to set the `COMPlus_GCHeapHardLimit` with these new changes, as it should be impossible for memory usage to runaway bad enough to warrant it.  Memory constrained systems may still want to use this option if they'd prefer that the application crashes itself before crashing the host.
